datafusion/expr/src/logical_plan/tree_node.rs

```diff
@@ -32,23 +39,362 @@ impl TreeNode for LogicalPlan {
+                    }),
+                    DdlStatement::CreateView(CreateView {
+                        name,
+                        input,
+                        or_replace,
+                        definition,
+                    }) => rewrite_arc(input, f)?.update_data(|input| {
+                        DdlStatement::CreateView(CreateView {
+                            name,
+                            input,
+                            or_replace,
+                            definition,
+                        })
+                    }),
+                    // no inputs in these statements
+                    DdlStatement::CreateExternalTable(_)
+                    | DdlStatement::CreateCatalogSchema(_)
+                    | DdlStatement::CreateCatalog(_)
+                    | DdlStatement::DropTable(_)
+                    | DdlStatement::DropView(_)
+                    | DdlStatement::DropCatalogSchema(_)
+                    | DdlStatement::CreateFunction(_)
+                    | DdlStatement::DropFunction(_) => Transformed::no(ddl),
+                }
+                .update_data(LogicalPlan::Ddl)
+            }
+            LogicalPlan::Unnest(Unnest {
+                input,
+                column,
+                schema,
+                options,
+            }) => rewrite_arc(input, f)?.update_data(|input| {
+                LogicalPlan::Unnest(Unnest {
+                    input,
+                    column,
+                    schema,
+                    options,
+                })
+            }),
+            LogicalPlan::Prepare(Prepare {
+                name,
+                data_types,
+                input,
+            }) => rewrite_arc(input, f)?.update_data(|input| {
+                LogicalPlan::Prepare(Prepare {
+                    name,
+                    data_types,
+                    input,
+                })
+            }),
+            LogicalPlan::RecursiveQuery(RecursiveQuery {
+                name,
+                static_term,
+                recursive_term,
+                is_distinct,
+            }) => map_until_stop_and_collect!(
+                rewrite_arc(static_term, &mut f),
+                recursive_term,
+                rewrite_arc(recursive_term, &mut f)
+            )?
+            .update_data(|(static_term, recursive_term)| {
+                LogicalPlan::RecursiveQuery(RecursiveQuery {
+                    name,
+                    static_term,
+                    recursive_term,
+                    is_distinct,
+                })
+            }),
+            // plans without inputs
+            LogicalPlan::TableScan { .. }
+            | LogicalPlan::Statement { .. }
+            | LogicalPlan::EmptyRelation { .. }
+            | LogicalPlan::Values { .. }
+            | LogicalPlan::DescribeTable(_) => Transformed::no(self),
+        })
     }
 }
+
+/// Converts a `Arc<LogicalPlan>` without copying, if possible. Copies the plan
```

Comment:
Here is the code that avoids copying for `Arc<LogicalPlan>` when possible (my performance results show it is possible most of the time)